enclave/src/polkadex_gateway.rs

```diff
@@ -114,42 +114,85 @@ impl PolkaDexGatewayCallback for PolkaDexGatewayCallbackImpl {
     }
 }
 
-/// Place order function does the following
-/// 1. authenticate
-/// 2. mutate balances (reserve amount offered in order)
-/// 3. store_order (async)
-/// 4. send order to OpenFinex API
-/// 5. report OpenFinex API result to sender
-pub fn place_order(
-    main_account: AccountId,
-    proxy_acc: Option<AccountId>,
-    order: Order,
-) -> Result<(), GatewayError> {
-    // Authentication
-    if let Err(e) = authenticate_user(main_account.clone(), proxy_acc) {
-        error!("Could not authenticate user due to: {:?}", e);
-        return Err(e);
-    };
-    // Mutate Balances
-    match order.order_type {
-        OrderType::LIMIT => {
-            if order.quantity == 0 as Balance {
-                error!("Limit Order quantity Zero");
-                return Err(GatewayError::QuantityZeroInLimitOrder);
+/// All sendings to the openfinex server should go through
+/// this gateway. Necessary to mock unit tests
+pub struct OpenfinexPolkaDexGateway<B: OpenFinexApi> {
+    openfinex_api: B
+}
```

Comment:
This is interesting, I didn't know you could compose a struct like this, using a trait as type parameter and then use it as trait object

Reply:
It was my first time using it like this aswell. I copied from the substrate code (top_pool has a lot of these kind of implementations: https://github.com/Polkadex-Substrate/polkadexTEE-worker/blob/master/enclave/src/top_pool/pool.rs#L130-L135)